### DISCUSSION TITLE
Should lookbook continue to support older versions of Ruby/Rails?

### DISCUSSION THREAD

**Discussion by DRBragg** [lookbook-hq/lookbook#199] [General]
Hey @allmarkedup, I wanted to post this here before opening an issue or PR. 

It looks like a lot of work recently went into how some of the backend and configuration for lookbook works. I haven't gone though all of it yet but I do know that some of the changes are incompatible with Ruby < 2.7. In the past I have opened PRs to keep lookbook (mostly) compatible with Rails 5.1+ and Ruby 2.5+ but I know these are very old versions at this point. I'm happy to keep opening issues and/or PRs to keep lookbook working with older Ruby/Rails but I understand if this isn't the direction you want to go. I just wanted to check before I went down that rabbit hole. If we need to pin to 1.1 until we're on newer Ruby/Rails that's totally fair.

Let me know what you think.

**Comment by allmarkedup** [comment]
Hey @DRBragg - I know this is something we've discussed before and I'm sorry that it's a constant headache for you.

As you may know the current Lookbook Github actions CI tests against Ruby 2.7+ and Rails 6.0+. Because I've so far failed to properly settle on and document an 'official" version support commitment that has served as the de-facto base level so far. But that CI situation is only really because when I set it up I was having a nightmare trying to get it set up to test against older versions or Ruby/Rails! As ViewComponent currently claims support for Ruby 2.5+ & Rails 5.0+ I think that Lookbook really should aim to support those versions too.

Before committing 100% to that however I'm going to try again to get Ruby 2.5 & Rails 5.x running to test against so that I can at least get a feel for how much work it'll be to fix the recent changes and then keep those older versions supported. I'm very new to Ruby so I definitely lack any clear idea of what language features are new and what have been around for ages. It's definitely a learning journey for me :-) Obviously all your PRs have been very helpful in keeping things working but without the tests running against those versions new releases are always likely to cause some headaches for folks in your situation.

You may also know that as of [v2.72](https://github.com/ViewComponent/view_component/releases/tag/v2.72.0) ViewComponent has deprecated support for Ruby < 2.7 with the intention of dropping support for it whenever ViewComponent v3.0 rolls around. I'll probably aim to do something similar to match the ViewComponent support requirements and release a new major version around that time.

But first I'll try to find a bit of time to get Lookbook running and tested against those older versions - I'll let you know as soon as I've made any progress.

**Reply by DRBragg** [reply]
Nothing at all to apologize for. It's unfair to expect any gem to continue to support EOL'ed versions of Ruby or Rails so we appreciate your willingness to accept my compatibility PRs. As always, I'm happy to help anyway I can I just didn't want to go down the rabbit hole if you weren't interested.

If it helps, I have [a gem](https://github.com/DRBragg/redaction) that tests against multiple versions of Ruby and Rails that you may be able to use to get something started for lookbook. 

Please let me know if there's anything I can do to help. I'm happy to open PRs, create issues, or pair on anything.  Thanks again for all you do, lookbook is awesome!

**Reply by allmarkedup** [reply]
Thanks @DRBragg, appreciate your offer of help on this - you've clearly got a lot more knowledge of keeping things running on older Ruby versions than I do!

Whilst 2.5 is EOL I think the fact that ViewComponent itself still officially supports it means Lookbook should too, if at all possible.

I've made a little progress today by finally managing to get Ruby 2.5 and Rails 5.2 added to the test matrix (and actually running!) in Github Actions. The tests are obviously failing currently but at least I've got a target to work towards now. Although I've failed completely so far to get Ruby 2.5 building on my M1 mac so local testing is a bit more of an issue still 🙄

I've opened a [draft PR](https://github.com/allmarkedup/lookbook/pull/201) with the CI setup changes and which I'll push Ruby 2.5-related fixes to. Then it can be merged into main when CI goes green. Hopefully there won't be too much that needs changing so I'll try to chip away at it when I can :-)

**Reply by DRBragg** [reply]
Looks like your current failure is related to this: https://github.com/allmarkedup/lookbook/blob/42e51d945a813d4de40065999dff7b09b6b89494/lib/lookbook/support/store.rb#L6

Numbered params for blocks wasn't added until 2.7

Simply changing it `initial_data.to_h.each { |k, v| self[k] = v }` should work fine.
